node/src/types/deploy/deploy_header.rs

```diff
@@ -91,13 +92,49 @@ impl DeployHeader {
         &self.chain_name
     }
 
-    /// Determine if this deploy header has valid values based on a `DeployConfig` and timestamp.
-    pub fn is_valid(&self, deploy_config: &DeployConfig, current_timestamp: Timestamp) -> bool {
-        let ttl_valid = self.ttl() <= deploy_config.max_ttl;
-        let timestamp_valid = self.timestamp() <= current_timestamp;
-        let not_expired = !self.expired(current_timestamp);
-        let num_deps_valid = self.dependencies().len() <= deploy_config.max_dependencies as usize;
-        ttl_valid && timestamp_valid && not_expired && num_deps_valid
+    /// Returns Ok if and only if the dependencies count and TTL are within limits, and the
+    /// timestamp is not later than `at`.  Does NOT check for expiry.
+    pub fn is_valid(
+        &self,
+        config: &DeployConfig,
+        at: Timestamp,
+        deploy_hash: &DeployHash,
+    ) -> Result<(), DeployConfigurationFailure> {
+        if self.dependencies.len() > config.max_dependencies as usize {
+            debug!(
+                %deploy_hash,
+                deploy_header = %self,
+                max_dependencies = %config.max_dependencies,
+                "deploy dependency ceiling exceeded"
+            );
+            return Err(DeployConfigurationFailure::ExcessiveDependencies {
+                max_dependencies: config.max_dependencies,
+                got: self.dependencies().len(),
+            });
+        }
+
+        if self.ttl() > config.max_ttl {
+            debug!(
+                %deploy_hash,
+                deploy_header = %self,
+                max_ttl = %config.max_ttl,
+                "deploy ttl excessive"
+            );
+            return Err(DeployConfigurationFailure::ExcessiveTimeToLive {
+                max_ttl: config.max_ttl,
+                got: self.ttl(),
+            });
+        }
+
+        if self.timestamp() > at {
```

Comment:
Shall we consider allowing a potential, small gap to compensate for the fact that clocks are never perfectly in sync? So I'd say it should be ok if the deploy timestamp is, let's say, 5 sec. in the future?

I think a similar gap is allowed in authenticator kind of apps.

Reply:
I guess the node's clock could be a bit behind the client. Though here `at` will be higher than `self.timestamp()` if the clocks are perfectly in sync because of the network delay for sending the deploy. But since we strive to have the clocks in sync, the difference would be on the order of tens for hundreds of miliseconds at most.
I would say we shouldn't add additional complexity here. Since the client wants to communicate with this node, it can account for the clock difference itself and delay the creation of the deploy by a few miliseconds if it receives feedback that the deploy is in the future.

Reply:
Yup, I agree with @alsrdn.